rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+improve the internals of each.
+
+### Subcomponents
+
+Based on the user-level concerns we collected from the list of issues, let's
+break down what each respective subcomponent should look like.
+
+#### Read scheduling
+
+In the pseudocode above, read scheduling is controlled by `sort`, `should_read`,
+`limit_reached`, `should_not_read_next_file`, and `maybe_backoff`. A component
+would need to answer the following questions:
+
+1. In what order should I read the available files?
+1. Should I finish one file before moving on to the next?
+1. Should I back off reads to this file?
+1. Should I back off reads to all files (i.e. sleep)?
+1. How long should I spend working on a single file?
+
+These questions would then be answered by a combination of configuration, file
+metadata, and gathered statistics.
+
+There are (at least) two approaches we could take to building this component.
+The most obvious would be to implement a struct that contains configuration and
+exposes methods very much like the ones mentioned above. This would maintain the
+current structure of one main read loop with multiple points of control. An
+alternative would be to introduce a trait representing the logic of the read
+loop. This may lead to some duplication, but would allow simpler separation of
+different use cases and require readers of the code understand the interaction
+between fewer subtle points of control.
+
+I would propose that we begin with the simpler method of consolidating logic
+from the existing structure. This involves fewer design decisions and we may
+find with the other simplifications we're planning make that further simplifying
+the read loop is not worthwhile. That being said, once related features (like
+the batch mode discussed later) are complete, we can reevaluate this decision.
+
+#### File identity
+
+Identity is an area where we already have the beginning of modularity with the
+`Fingerprinter` abstraction, but it's incomplete. In our pseudocode, identity is
+involved in both `look_for_files` and `reconcile`. It's not just a way to figure
+out a magic identifier, but also the logic to update our list of watched files
+based on those identifiers. It needs to answer the following:
+
+1. Given a visible path, does it contain a file I've seen before?
+1. If I have seen it before, has it been renamed?
+1. If I have seen it before, am I now seeing it in multiple places?
+1. If I'm seeing duplicates, how should I choose which to follow?
+
+This logic is mostly in one place in the current implementation, so it should
+not be terribly difficult to extract it. The use of `Fingerprinter` should
+likely become purely internal to the new component.
+
+In addition to simply consolidating the logic, we can expand and make the use of
+`Fingerprinter` more intelligent. We currently have three ways it can work:
+
+1. Checksum (usually reliable but frustrating for small files)
+1. Device and inode (simple and works with small files, but doesn't handle edge
```

Comment:
Which edge cases?

Reply:
[Inode reuse](https://www.elastic.co/guide/en/beats/filebeat/current/inode-reuse-issue.html) is one example. We've also had reports of it having unreliable behavior with Docker volume mounts, etc.